### ISSUE TITLE
Extend in-memory storage to support Integers other than 64-bit, and fixed-length/length-capped strings

### ISSUE THREAD

**Issue by wdanilo** [enso-org/enso#5159]
This task is automatically imported from the old Task Issue Board and it was originally created by James Dunkerley.
Original issue is [here](https://www.pivotaltracker.com/story/show/183854206).
<hr />

Single place to manage the type coercion of a Value_Type.

## Updated spec

Our storage currently only supports 64-bit integers and variable-length strings.

Add a way to parametrize the storage/builder and make it accept only smaller numbers if asked.
Similarly, our text storage needs to take an optional maximum length and a flag specifying if the texts stored are fixed-length or variable-length.

This is needed to be able to use `Table.cast` with in-memory columns properly.

- [ ] Extend storages with additional metadata.
- [ ] If an integer is too big for the requested datatype, report a `Conversion Failure` and replace it with `Nothing`.
	- Once non-nullable types are implemented we may need some other placeholder, or just to fail hard.
- [ ] If Text is too long, trunacte it (and report `Conversion Failure`).
- [ ] If Text is too short, pad it with spaces on the right.
- [ ] Ensure that tests for `Table.cast` are updated to cover the new scenarios.
- [ ] (If implemented after #6111) Ensure that tests for `Column.from_vector`, `Column.map` and `Column.zip` are updated to cover the new scenarios.
- [ ] Ensure that `Column|Table.parse` check integer ranges when parsing.
- [ ] _Update repros of #6270, #6258, #6257 because after this change the current repros will stop working, but that has nothing to do with the underlying issue - we just need to change the cast from `Char 1` which will now be supported to a `Float 32-bit`._

#### Non-goals

Currently we are still storing integers in 64-bit values, just 'fake' adding limits for the size of the values.

It is not a goal of this task to extend our storage system to store the smaller integers more compactly.  This is of course possible, but with the poor support for generics on primitive types in Java, will require lots of boilerplate. We probably would need to create a code generator to do it efficiently. Tracked in a separate, lower priority task: #6109

Since we cannot limit precision of floating point numbers to 32-bit float when storing them in 64-bit storage (since the layout is different), we are not supporting 32-bit floats for now. We can add it as part of the task above.

**Comment by radeusgd** [comment]
~~Update the `merge_type` method in SQL_Type to work on Value Type~~

**Comment by radeusgd** [comment]
Some things to take into account when implementing fixed-length string storage.

What strings fit into the type `Value_Type.Char size=n` may highly depend on the database.

```
from Standard.Base import all
from Standard.Table import all
from Standard.Database import all

main =
    #c = Database.connect (SQLite In_Memory)
    c = connect_to_postgres

    t0 = Table.new [["X", ['abśxy', 'abs\u0301xy', 'ab🤦‍♂️xy']]]
    t0.print

    cx = t0.at "X"
    IO.println (cx.map .length . to_vector)
    IO.println ((cx.at 0) == (cx.at 1))

    # We don't have limited column length in-memory yet so best we can do is use our basic Text functions
    IO.println (cx.map (s-> s.take 3) . to_vector)

    t1 = t0.select_into_database_table c "t1" temporary=True
    t1.print

    t2 = t1.cast "X" (Value_Type.Char size=3)
    IO.println "Warnings?: "+(Warning.get_all t2 . map .value).to_display_text
    t2.at "X" . to_vector


connect_to_postgres =
    hostname = Environment.get "ENSO_DATABASE_TEST_HOST"
    db_name = Environment.get "ENSO_DATABASE_TEST_DB_NAME"
    db_host_port = hostname.if_nothing "localhost" . split ':'
    db_host = db_host_port.at 0
    db_port = if db_host_port.length == 1 then 5432 else Integer.parse (db_host_port.at 1)
    db_user = Environment.get "ENSO_DATABASE_TEST_DB_USER"
    db_password = Environment.get "ENSO_DATABASE_TEST_DB_PASSWORD"
    case db_name.is_nothing of
        True ->
            IO.println "PostgreSQL test database is not configured. See README.md for instructions."
        False ->
            connection = Panic.rethrow <|
                IO.println "Connecting to PostgreSQL test database at "+hostname
                Database.connect (Postgres db_host db_port db_name credentials=(Credentials.Username_And_Password db_user db_password))
            connection
 ```
gives
```
Connecting to PostgreSQL test database at 172.28.138.202
   | X
---+-------
 0 | abśxy
 1 | abśxy
 2 | ab🤦‍♂️xy

[5, 5, 5]
True
[abś, abś, ab🤦‍♂️]
 X
-------
 abśxy
 abśxy
 ab🤦‍♂️xy

Warnings?: []
['abś', 'abs', 'ab🤦']
```

This shows that if we rely on Enso string length (based on grapheme clusters), we will get different results in-memory than in Postgres. 

That may be desirable - after all - we want the in-memory backend to be consistent with Enso operations.

Alternatively, we could try to be more consistent with databases - e.g. to limit the length by count of code units. Yet, this may still not be consistent, for example let's take the query `SELECT CAST('ąbcd' AS CHAR(3)) AS "C";`, in a Postgres DB with `UTF-8` server side encoding, it will 'correctly' yield `ąbc` as result, however if the encoding is `SQL_ASCII`, it will return just `ąb` (because `SQL_ASCII` essentially looks at bytes and `ą` takes two bytes to encode in the client-side UTF-8 encoding). However, the `SQL_ASCII` encoding seems to be rather an outlier as it can do even worse - if we do `SELECT CAST('ąąą' AS CHAR(3)) AS "C";` it will actually fail with `error: invalid byte sequence for encoding "UTF8": 0xc4` - the `SQL_ASCII` encoding breaks the string in the middle, making it impossible to decode back in UTF-8.

I see two possibilities for the interpretation of `size` in `Value_Type.Char`:
1. Interpret it as a Java string length, hoping that it will be mostly compatible with how databases will handle this as well.
2. Interpret it using the 'native'/'default' notion for each backend - i.e. relying on Enso grapheme-cluster-based length for the in-memory backend.

In most default cases both will be indistinguishable. The differences only show up when working with grapheme clusters that take multiple code points, like emojis or in some more complicated scripts (i.e. Sanskrit).

I'm leaning towards picking (2) as both have some + and - and being consistent in-memory with Enso seems most valuable (especially as even if we pick (1) we still don't get 100% consistency, because we may encounter 'exotic' databases that will still do differently).

@jdunkerley does (2) sound good?
